statsmodels/tsa/statespace/mlemodel.py

```diff
@@ -282,6 +300,150 @@ def loglike(self, params=None, average_loglike=False, transformed=True,
+
+        """
+        if params is not None:
+            self.update(params, transformed=transformed, set_params=set_params)
+
+        return super(MLEModel, self).loglikeobs(**kwargs)
+
+    def observed_information_matrix(self, params, **kwargs):
+        """
+        Observed information matrix
+
+        Parameters
+        ----------
+        params : array_like, optional
+            Array of parameters at which to evaluate the loglikelihood
+            function.
+        **kwargs
+            Additional keyword arguments to pass to the Kalman filter. See
+            `KalmanFilter.filter` for more details.
+
+        Notes
+        -----
+        This method is from Harvey (1989), which shows that the information
+        matrix only depends on terms from the gradient. This implementation is
+        partially analytic and partially numeric approximation, therefore,
+        because it uses the analytic formula for the information matrix, with
+        numerically computed elements of the gradient.
+
+        References
+        ----------
+        Harvey, Andrew C. 1990.
+        Forecasting, Structural Time Series Models and the Kalman Filter.
+        Cambridge University Press.
+
+        """
+        # Setup
+        n = len(params)
+        epsilon = _get_epsilon(params, 1, None, n)
+        increments = np.identity(n) * 1j * epsilon
+
+        # Get values at the params themselves
+        self.update(params)
+        res = self.filter(**kwargs)
+        dtype = self.dtype
+        # Save this for inversion later
+        inv_forecasts_error_cov = res.forecasts_error_cov.copy()
+
+        # Compute partial derivatives
+        partials_forecasts_error = (
+            np.zeros((self.k_endog, self.nobs, n))
+        )
+        partials_forecasts_error_cov = (
+            np.zeros((self.k_endog, self.k_endog, self.nobs, n))
+        )
+        for i, ih in enumerate(increments):
+            self.update(params + ih)
+            res = self.filter(**kwargs)
+
+            partials_forecasts_error[:, :, i] = (
+                res.forecasts_error.imag / epsilon[i]
+            )
+
+            partials_forecasts_error_cov[:, :, :, i] = (
+                res.forecasts_error_cov.imag / epsilon[i]
+            )
+
+        # Compute the information matrix
+        tmp = np.zeros((self.k_endog, self.k_endog, self.nobs, n), dtype=dtype)
+
+        information_matrix = np.zeros((n, n), dtype=dtype)
+        for t in range(self.loglikelihood_burn, self.nobs):
+            inv_forecasts_error_cov[:, :, t] = (
+                np.linalg.inv(inv_forecasts_error_cov[:, :, t])
+            )
+            for i in range(n):
+                tmp[:, :, t, i] = np.dot(
+                    inv_forecasts_error_cov[:, :, t],
+                    partials_forecasts_error_cov[:, :, t, i]
+                )
+            for i in range(n):
+                for j in range(n):
```

Comment:
Three levels of nested loops, but I'm not sure there's anything to be done about it.